subprojects/core/src/main/java/org/gradle/process/internal/DefaultSystemPropertyHandler.java

```diff
@@ -0,0 +1,180 @@
+
+import com.google.common.collect.ImmutableList;
+import org.gradle.api.Named;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputDirectory;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.Internal;
+import org.gradle.api.tasks.Nested;
+import org.gradle.api.tasks.OutputDirectory;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.PathSensitive;
+import org.gradle.api.tasks.PathSensitivity;
+import org.gradle.process.CommandLineArgumentProvider;
+import org.gradle.process.InputOutputValue;
+import org.gradle.process.JavaForkOptions;
+import org.gradle.process.SystemPropertyHandler;
+import org.gradle.util.DeferredUtil;
+
+public class DefaultSystemPropertyHandler implements SystemPropertyHandler {
+    private final JavaForkOptions javaForkOptions;
+
+    public DefaultSystemPropertyHandler(JavaForkOptions javaForkOptions) {
+        this.javaForkOptions = javaForkOptions;
+    }
+
+    @Override
+    public void add(String name, InputOutputValue value) {
+        javaForkOptions.getJvmArgumentProviders().add(new SystemPropertyCommandLineArgument(name, value));
+    }
+
+    @Override
+    public void add(String name, Object value) {
+        javaForkOptions.systemProperty(name, value);
+    }
+
+    @Override
+    public InputOutputValue inputFile(Object file) {
+        return new InputFileValue(file);
+    }
+
+    @Override
+    public InputOutputValue inputDirectory(Object file) {
+        return new InputDirectoryValue(file);
+    }
+
+    @Override
+    public InputOutputValue outputFile(Object file) {
+        return new OutputFileValue(file);
+    }
+
+    @Override
+    public InputOutputValue outputDirectory(Object file) {
+        return new OutputDirectoryValue(file);
+    }
+
+    @Override
+    public InputOutputValue ignored(Object file) {
+        return new IgnoredValue(file);
+    }
+
+    private class SystemPropertyCommandLineArgument implements CommandLineArgumentProvider, Named {
+        private final String name;
+        private final InputOutputValue inputOutputValue;
+
+        public SystemPropertyCommandLineArgument(String name, InputOutputValue inputOutputValue) {
+            this.name = name;
+            this.inputOutputValue = inputOutputValue;
+        }
+
+        @Nested
+        public InputOutputValue getValue() {
+            return inputOutputValue;
+        }
+
+        @Input
+        @Override
+        public String getName() {
+            return name;
+        }
+
+        @Override
+        public Iterable<String> asArguments() {
+            return ImmutableList.of("-D" + name + "=" + DeferredUtil.unpack(inputOutputValue.getValue()));
+        }
+    }
+
+    private class BaseInputOutputValue implements InputOutputValue {
+
+        private final Object value;
+
+        public BaseInputOutputValue(Object value) {
+            this.value = value;
+        }
+
+        @Override
+        public Object getValue() {
```

Comment:
It is okay to declare `getValue()` here I think, and just use `BaseIOV` instead of `IOV` internally.

Reply:
Or we could have an internal interface and just re-implement storing the actual object multiple times.

Reply:
This thing also sounds a lot like a `Provider` in some way...

Reply:
I refactored the code.